### ISSUE TITLE
Performance issue with response time when upgrading from 3.31.0 to 3.31.2

### ISSUE THREAD

**Issue by op-rasmusl** [Azure/azure-cosmos-dotnet-v3#3613]
**Describe the bug**
Increased response time since updating from 3.31.0 to 3.31.2. This impacted around 5% of http requests. 

**To Reproduce**
When we updated from Microsoft.Azure.Cosmos 3.31.0 to 3.31.2 we noticed an increase in response time for our http requests. We use a single cosmos client in our services. 

The code that talks to cosmos is a cross partition query, see below:
```
public async Task<IEnumerable<T>> GetAll(Expression<Func<T, bool>> predicate)
{
    var iterator = container.GetItemLinqQueryable<T>()
        .Where(entity => entity.Kind == DocumentKind)
        .Where(predicate)
        .ToFeedIterator();

    var entities = new List<T>();

    while (iterator.HasMoreResults)
    {
        var page = await iterator.ReadNextAsync();
        entities.AddRange(page);
    }

    return entities;
}
```
The function is used as follows:
```
GetAll(item => item.Key == value)
```

The container has roughly 5 million items and is partioned. 

**Expected behavior**
We expect the upgrade to not impact performance. 

**Actual behavior**
An increase in response time for http requests. 

<img width="1921" alt="github_graph" src="https://user-images.githubusercontent.com/112619777/208648173-e0981710-159e-4d03-8727-c1b551c08a6b.png">

The graph shows the response time in milliseconds for one of our endpoints, for the 99th percentile. In the 50th percentile there is no noticeable difference. The peaks are with the sdk version 3.31.2 and when we downgraded to 3.31.0 the response time returned to its usual duration. The second peak is when we upgraded to 3.31.2 again to see if that was the issue. This has been verified in both .NET 6 and .NET 7. The only change between the peaks is upgrading/downgrading Microsoft.Azure.Cosmos. 

**Environment summary**
SDK Version: .NET 7.0.1 and .NET 6.0.12
OS Version debian.11-x64

**Comment by ealsur** [comment]
@op-rasmusl Can you share the Diagnostics (https://docs.microsoft.com/en-us/azure/cosmos-db/sql/troubleshoot-dot-net-sdk-slow-request#capture-diagnostics) for both 3.31.0 and 3.31.2?

When I check the changelog (https://github.com/Azure/azure-cosmos-dotnet-v3/blob/master/changelog.md#-3312---2022-11-03) there are only 2 changes, one is related to connection health, the other to queries (but for ORDER BY), I don't see any of them affecting latency on P99 because the query you showed has no ORDER BY.

Do you perform other operations besides these?
Is the P99 calculated only for queries? Is it for all queries regardless of the predicate?

**Comment by ealsur** [comment]
> The orange line is requests to {cosmos account name}.documents.azure.com
> The blue line is requests to {cosmos account name}-westeurope.documents.azure.com

But you are just measuring dependency calls, how do you know those dependency calls are related to those 2 operations you are performing and not others?

The SDK does perform background operations (like prefetching results, or refreshing the account information) that do not impact the actual operation being executed.

My understanding of Application Insights is that, unless you are implementing your own tracking, it will automatically measure HTTP requests, but are you measuring the latency of any of these 2 operations particularly? For example, using `Diagnostics.GetClientElapsedTime()` or through some Stopwatch mechanism? If these are raw HTTP requests, the increase in latency would not make sense to be related to the Cosmos SDK, once the request is sent, the library cannot control the latency on the network.

**Comment by op-rasmusl** [comment]
We agree that this seems farfetched but we are not sure what else it could be. You can close this issue if you want. We will do a deeper investigation with diagnostics and get back to you if we find anything relevant.

**Comment by ealsur** [comment]
@op-rasmusl let's leave it open for the time being. I would be interested in the Diagnostics to understand if this is merely background HTTP operations or your actual SDK calls are getting affected.

If you are not changing the Connection Mode (the default is TCP), keep in mind that Data Plane calls (like ReadItemAsync) are performed through TCP and do not appear in Application Insights (hence the question of if this is really affecting your operational latency).

**Comment by op-rasmusl** [comment]
Hello again, 

By only updating from 3.31.0 to 3.31.1 we managed to reproduce the issue, which rules out 3.31.2 as the cause. We have also run diagnostics on the cosmos reponses and can share the logs, however we are not sure if there is any sensitive data in them. How do we send these to you or are they safe to share on github?

**Comment by ealsur** [comment]
The diagnostics will contain just the name of your account, you can just remove that or replace it with a placeholder.

Also, going back to previous questions: Are you still looking at dependencies? Did you get to measure the actual operation time (either by your own stopwatch or using `Diagnostics.GetClientElapsedTime()`) and see a change in the P99 across these versions?

**Comment by op-rasmusl** [comment]
> But you are just measuring dependency calls, how do you know those dependency calls are related to those 2 operations you are performing and not others?

We know they are related because we looked at the call stack in application insights for a selection of requests that are in the P99. Just looking at the P99 duration for dependencies shows the same pattern. 

The logs we are sending were collected from our cosmos responses where ```Diagnostics.GetClientElapsedTime()``` was larger than 800ms. In 3.31.0 we rarely have such high reponse times however with 3.31.1 we get these regularely. 

[cosmos-diagnostics.zip](https://github.com/Azure/azure-cosmos-dotnet-v3/files/10350582/cosmos-diagnostics.zip)

**Comment by ealsur** [comment]
`cosmos_diagnostics.json` contains HTTP timeouts, that is the reason of the latency of that operation, the SDK retried the timeout.

![image](https://user-images.githubusercontent.com/1633401/210837258-e58e1b7d-12a8-4f20-bab3-4f478f6066ea.png)

`cosmos_diagnostics2.json` looks like it's the first request to a Container, reference https://learn.microsoft.com/en-us/azure/cosmos-db/nosql/sdk-connection-modes#routing. If your app is working on multiple Containers, you will see latency spikes on the first operations for each Container.

![image](https://user-images.githubusercontent.com/1633401/210837434-6b7b4430-be68-4922-840e-473b465ac5e4.png)

`cosmos_diagnostics3.json` has network latency, we see 900ms on the network wire:

![image](https://user-images.githubusercontent.com/1633401/210837760-dd2e547d-2aa8-42f8-82c5-6851b46fa2b9.png)

`cosmos_diagnostics4.json` has high Received time (reference https://learn.microsoft.com/en-us/azure/cosmos-db/nosql/troubleshoot-dotnet-sdk-slow-request?tabs=cpu-new#rntbdrequeststats). Do you have any place in the code where you might be blocking async calls?

![image](https://user-images.githubusercontent.com/1633401/210838248-ee751478-9cdb-404d-9306-6bb72c373d83.png)

`cosmos_diagnostics5.json` has network latency, similar to `cosmos_diagnostics3.json`:

![image](https://user-images.githubusercontent.com/1633401/210838748-4d39e71d-59e7-450c-8575-2f2cd116092d.png)

None of these cases seem to be related to any change introduced in 3.31.0. 

From the App Insights perspective (that traces HTTP calls only), the only that would appear there is `cosmos_diagnostics2.json`, that contains the Read Collection call (HTTP), the others (high time in Transit Time which is TCP, or high time in Received which is TCP, or `cosmos_diagnostics1.json` with the timeout) would not appear in dependency latency.

**Comment by op-rasmusl** [comment]
Thank you for the feedback on the logs, it seems that for now it is still unclear where the problem stems from. We will continue to investigate this, however you can close this issue if you want.

**Comment by ealsur** [comment]
Cleaning up issues with needs-investigation that are dormant/not-actionable

**Comment by op-rasmusl** [comment]
We have been working on this issue for a while and have a few observations:

1. We have reviewed our code and have not found any suspicious async code
2. We have changed networking and are now using Private Endpoints between the App Service and Cosmos 

We have tried to update to 3.31.1 and above, however the 99 problem still occurs. What did solve it was updating from the P1V2 app service plan to the P1V3 app service plan which allowed us to update the sdk without any performance issues. The main difference seems to be that we now have two cores on each instance. We are not sure why this matters since we rarely had a high cpu percentage and did not to have threading issues.
